src/Token/TokenProvider.php

```diff
@@ -0,0 +1,279 @@
+    public static function fromToken(TokenInterface $token)
+    {
+        $promise = Promise\Create::promiseFor($token);
+
+        return function () use ($promise) {
+            return $promise;
+        };
+    }
+
+    /**
+     * Creates an aggregate token provider that invokes the provided
+     * variadic providers one after the other until a provider returns
+     * a token.
+     *
+     * @return callable
+     */
+    public static function chain()
+    {
+        $links = func_get_args();
+        //Common use case for when aws_shared_config_files is false
+        if (empty($links)) {
+            return function () {
+                return Promise\Create::promiseFor(false);
+            };
+        }
+
+        return function ($previousToken = null) use ($links) {
+            /** @var callable $parent */
+            $parent = array_shift($links);
+            $promise = $parent();
+            while ($next = array_shift($links)) {
+                if ($next instanceof RefreshableTokenProviderInterface
+                    && $previousToken instanceof Token
+                ) {
+                    $promise = $promise->otherwise(
+                        function () use ($next, $previousToken) {return $next->refresh($previousToken);}
+                    );
+                } else {
+                    $promise = $promise->otherwise($next);
+                }
+            }
+            return $promise;
+        };
+    }
+
+    /**
+     * Wraps a token provider and caches a previously provided token.
+     *
+     * Ensures that cached tokens are refreshed when they expire.
+     *
+     * @param callable $provider Token provider function to wrap.
+     *
+     * @return callable
+     */
+    public static function memoize(callable $provider)
+    {
+        return function () use ($provider) {
+            static $result;
+            static $isConstant;
+
+            // Constant tokens will be returned constantly.
+            if ($isConstant) {
+                return $result;
+            }
+
+            // Create the initial promise that will be used as the cached value
+            // until it expires.
+            if (null === $result) {
+                $result = $provider();
+            }
+
+            // Return a token that could expire and refresh when needed.
+            return $result
+                ->then(function (TokenInterface $token) use ($provider, &$isConstant, &$result) {
+                    // Determine if the token is constant.
+                    if (!$token->getExpiration()) {
+                        $isConstant = true;
+                        return $token;
+                    }
+
+                    // Refresh a token in the expiration window
+                    if (!$token->isExpired()) {
+                        return $token;
+                    }
+                    // Refresh the result and forward the promise.
+                    return $result = $provider($token);
```

Comment:
This comment spans a few different places: I don't think it's entirely necessary for us to pass the previous credentials around.  If this memoize method works like the one in the credential provider class, the `$token` here should be the current token, which in that case you could just invoke the provider again. We definitely want to keep the refresh logic (within ~10 min of expiration) though. I haven't pulled this down to test quite yet so I might be missing something here.